### ISSUE TITLE
Problems about kvazaar Coding speed on development board

### ISSUE THREAD

**Issue by YanjunHH** [ultravideo/kvazaar#213]
Hello. 
We are doing research on the real-time performance of kvazaar encoder. We hope to achieve 1080P real-time encoding effect (30FPS) on ZYNQ UltraScale + ZCU102 Evaluation Platform .The version of the kvazaar encoder is 1.2.0. But the encoding speed is only about 1FPS now. We use the vivado HLS tool to synthesize the encoder, but we have encountered many problems.We can not synthesize the IP core.
How should we speed up the encoder?Can you provide some technical support or IP core for us to try?
Grateful！

**Comment by fador** [comment]
Hi,

The performance sounds similar that we have encountered in Raspberry Pi 3, containing a similar ARM prosessor setup. One reason is that we have not optimized Kvazaar for ARM NEON and because the performance of those processors is not comparable to x86 based CPUs.

We have used HLS to accelerate Kvazaar, but using the version in Github directly will not work. I have not done HLS myself, but what I understand is that the memory access patterns and the undefined dataflow model make this unusable in an FPGA. Instead, we have developed a separate accelerator cores (currently for only Intra coding) and run the actual Kvazaar in an x86 based Xeon CPU while communicating over PCIe with the FPGA accelerator.

We mainly work with Altera FPGA boards and use Catapult C as the HLS tool.

Some references:
[FPGA-Powered 4K120p HEVC Intra Encoder](https://ieeexplore.ieee.org/document/8351873)
[High-level synthesized 2-D **IDCT/IDST** implementation for HEVC codecs on FPGA](https://ieeexplore.ieee.org/document/8050323)
[High-level synthesis implementation of HEVC 2-D **DCT/DST** on FPGA](https://ieeexplore.ieee.org/document/7952416)
[High-Level Synthesis Design Flow for HEVC Intra Encoder on SoC-FPGA](https://ieeexplore.ieee.org/document/7302250)

So in conclusion, I'm sorry to say it needs quite a lot of extra work to actually get Kvazaar running using HLS. We might be able to help with some problems, but even we do not have the full Kvazaar in an FPGA.

**Comment by YanjunHH** [comment]
Thank you for your reply and suggestions, we will continue to pay attention to the kvazaar encoder.

**Comment by YanjunHH** [comment]
Hi,
Have you tested the encoding speed of the kvazaar encoder running entirely on the Xeon CPU? 720P and 1080P video. We want to know how much the Intra coding accelerator cores improves the overall speed.
Looking forward to your reply!

**Comment by fador** [comment]
Hi,

We have results with 4k in our paper, with sequences available in our website:
[http://ultravideo.cs.tut.fi#testsequences](http://ultravideo.cs.tut.fi/#testsequences)
![](http://fador.be/g/3cf47bda3b8c2662f1bb90ff067b8cf1.png)

So in 4k we can get from 18fps -> 123fps on average, 6.8x speedup
I asked from the HLS/FPGA team and it seems that the 1080p results should be around 4x
Meaning 1080p should go from 70fps -> 490fps in this case.

720p should be around 2x from 1080p, but we didn't actually try to run any 720p sequences with this setup.

**Comment by YanjunHH** [comment]
Thank you for your reply, I am very grateful！

**Comment by YanjunHH** [comment]
Thank you for your reply, I am very grateful！

-----原始邮件-----
发件人:"Marko Viitanen" <notifications@github.com>
发送时间:2018-11-09 14:38:54 (星期五)
收件人: ultravideo/kvazaar <kvazaar@noreply.github.com>
抄送: YanjunHH <huangyanjun@pku.edu.cn>, Author <author@noreply.github.com>
主题: Re: [ultravideo/kvazaar] Problems about kvazaar Coding speed on development board (#213)



Hi,

We have results with 4k in our paper, with sequences available in our website:
http://ultravideo.cs.tut.fi#testsequences


So in 4k we can get from 18fps -> 123fps on average, 6.8x speedup
I asked from the HLS/FPGA team and it seems that the 1080p results should be around 4x
Meaning 1080p should go from 70fps -> 490fps in this case.

720p should be around 2x from 1080p, but we didn't actually try to run any 720p sequences with this setup.
